packages/core/src/routes/experience/classes/experience-interaction.ts

```diff
@@ -0,0 +1,164 @@
+});
+
+/**
+ * Interaction is a short-lived session session that is initiated when a user starts an interaction flow with the Logto platform.
+ * This class is used to manage all the interaction data and status.
+ *
+ * @see {@link https://github.com/logto-io/rfcs | Logto RFCs} for more information about RFC 0004.
+ */
+export default class ExperienceInteraction {
+  /**
+   * Factory method to create a new `ExperienceInteraction` using the current context.
+   */
+  static async create(ctx: WithLogContext, tenant: TenantContext) {
+    const { provider } = tenant;
+    const interactionDetails = await provider.interactionDetails(ctx.req, ctx.res);
+    return new ExperienceInteraction(ctx, tenant, interactionDetails);
+  }
+
+  /** The interaction event for the current interaction. */
+  private interactionEvent?: InteractionEvent;
+  /** The user verification record list for the current interaction. */
+  private readonly verificationRecords: Map<VerificationType, VerificationRecord>;
+  /** The accountId of the user for the current interaction. Only available once the user is identified. */
+  private accountId?: string;
+  /** The user provided profile data in the current interaction that needs to be stored to database. */
+  private readonly profile?: Record<string, unknown>; // TODO: Fix the type
+
+  constructor(
+    private readonly ctx: WithLogContext,
+    private readonly tenant: TenantContext,
+    interactionDetails: Interaction
+  ) {
+    const { libraries, queries } = tenant;
+
+    const result = interactionStorageGuard.safeParse(interactionDetails.result ?? {});
+
+    assertThat(
+      result.success,
+      new RequestError({ code: 'session.interaction_not_found', status: 404 })
+    );
+
+    const { verificationRecords = [], profile, accountId, event } = result.data;
+
+    this.interactionEvent = event;
+    this.accountId = accountId; // TODO: @simeng-li replace with userId
+    this.profile = profile;
+
+    this.verificationRecords = new Map();
+
+    for (const record of verificationRecords) {
+      const instance = buildVerificationRecord(libraries, queries, record);
+      this.verificationRecords.set(instance.type, instance);
+    }
+  }
+
+  /** Set the interaction event for the current interaction */
+  public setInteractionEvent(event: InteractionEvent) {
+    // TODO: conflict event check (e.g. reset password session can't be used for sign in)
+    this.interactionEvent = event;
+  }
+
+  /** Set the verified `accountId` of the current interaction from the verification record */
+  public identifyUser(verificationRecord: VerificationRecord) {
+    // Throws an 404 error if the user is not found by the given verification record
+    assertThat(
+      verificationRecord.verifiedUserId,
+      new RequestError({
+        code: 'user.user_not_exist',
+        status: 404,
+      })
+    );
+
+    // Throws an 409 error if the current session has already identified a different user
```

Comment:
nit: should we throw error once `this.accountId` has value? since "identify" is a state-mutating operation, while "Identified" is a non-mutable state.